torch/csrc/jit/runtime/symbolic_shape_registry.cpp

```diff
@@ -160,26 +163,130 @@ const at::optional<const FunctionSchema*> getInplaceVariant(
   return at::nullopt;
 }
 
-void registerSchema(
-    const FunctionSchema* schema_string,
-    const std::string& shape_compute_function_name,
-    std::unordered_map<std::string, std::shared_ptr<Graph>>& reused_functions,
-    const CompilationUnit& module) {
-  if (reused_functions.count(shape_compute_function_name)) {
-    auto graph = reused_functions[shape_compute_function_name];
+TypePtr mapTensorToListOfInts(TypePtr type) {
+  if (type->cast<TensorType>()) {
+    return ListType::ofInts();
+  }
+  at::ArrayRef<TypePtr> contained = type->containedTypes();
+  if (contained.empty()) {
+    return type;
+  }
+  return type->withContained(
+      fmap(type->containedTypes(), mapTensorToListOfInts));
+}
 
-    // allow extra unused arguments to map multiple functions to e.g. unary
+void checkForWhileLoop(
+    const FunctionSchema* schema,
+    std::shared_ptr<Graph> graph) {
+  DepthFirstGraphNodeIterator graph_it(graph);
+  for (auto* node = graph_it.next(); node != nullptr; node = graph_it.next()) {
+    if (node->kind() != prim::Loop) {
+      continue;
+    }
+    LoopView loop(node);
+    if (loop.loopType() != LoopView::For) {
+      TORCH_WARN(
+          "While loops are not yet implemented in unrolling which may make this shape function difficult to partially evaluate: ",
+          *node,
+          " for schema ",
+          *schema);
+    }
+  }
+}
+
+void checkInputReturnedAsOutput(
+    const FunctionSchema* schema,
+    const std::shared_ptr<Graph>& graph) {
+  // Could use alias db here as well but would have to warn because it's
```

Comment:
Nice. I think a few of my handwritten shape functions are breaking this invariant! Will try to upstream them and get this great linting :)

Reply:
The main thing I'm trying to preserve is that you can compose shape functions together... Technically it's fine if an aliased list is return so long as no shape function mutates its' input, however I thought it would be a little more defensive/safe to return a new unaliased list..